src/Facility.Core/Http/HttpClientService.cs

```diff
@@ -101,8 +103,66 @@ protected HttpClientService(HttpClientServiceSettings? settings, Uri? defaultBas
 			{
 				var contentType = mapping.RequestBodyContentType ?? requestHeaders?.GetContentType();
 				httpRequest.Content = GetHttpContentSerializer(requestBody.GetType()).CreateHttpContent(requestBody, contentType);
-				if (m_disableChunkedTransfer)
+				if (m_shouldCompressRequest(request))
+					httpRequest.Content = await CompressContentAsync(httpRequest.Content, cancellationToken).ConfigureAwait(false);
+				else if (m_disableChunkedTransfer)
 					await httpRequest.Content.LoadIntoBufferAsync().ConfigureAwait(false);
+
+				async static Task<HttpContent> CompressContentAsync(HttpContent httpContent, CancellationToken cancellationToken)
+				{
+					Stream? contentStream = null;
+					Stream? compressedContentStream = null;
+
+					try
+					{
+						// copy the existing HTTP content into a memory stream
+						contentStream = new MemoryStream();
+#if NET5_0_OR_GREATER
+						await httpContent.CopyToAsync(contentStream, cancellationToken).ConfigureAwait(false);
+#else
+						await httpContent.CopyToAsync(contentStream).ConfigureAwait(false);
+#endif
+						contentStream.Position = 0;
+
+						// compress that memory stream with gzip
+						compressedContentStream = new MemoryStream();
+						using (var gzipStream = new GZipStream(compressedContentStream, CompressionLevel.Optimal, leaveOpen: true))
+						using (var bufferedStream = new BufferedStream(gzipStream, 8192))
+						{
+#pragma warning disable CA1849 // Deliberately using synchronous copy for CPU-bound operation
+							contentStream.CopyTo(bufferedStream);
+#pragma warning restore CA1849 // Call async methods when in an async method
+						}
+
+						contentStream.Position = 0;
+						compressedContentStream.Position = 0;
+
+						// use the compressed result if it's fewer bytes to send
+						var useCompressedContent = compressedContentStream.Length < contentStream.Length;
+
+						// always construct a new HttpContent object (because we consumed the Stream from the original one)
+						var newHttpContent = new StreamContent(useCompressedContent ? compressedContentStream : contentStream);
+						if (useCompressedContent)
+							compressedContentStream = null;
+						else
+							contentStream = null;
+
+						// copy the headers from the original HttpContent object, adding Content-Encoding if compressed
+						foreach (var header in httpContent.Headers)
+							newHttpContent.Headers.TryAddWithoutValidation(header.Key, header.Value);
+						if (useCompressedContent)
+							newHttpContent.Headers.ContentEncoding.Add("gzip");
+
+						return newHttpContent;
+					}
+					finally
+					{
+#pragma warning disable CA1849 // No need to asynchronously dispose a MemoryStream
+						contentStream?.Dispose();
+						compressedContentStream?.Dispose();
```

Comment:
Do they need to be disposed at all?

Reply:
Technically, no, but it's the "polite" thing to do and helps automated tools like dotMemory that look for non-disposed objects (as a means of finding potential performance / leak bugs).